File: throwcode_python/algorithm/sort/quick_sort.py
#!/use/bin/env/python3
# -*- coding:utf-8 -*-
# @author: goinhn
# @date: 2021/3/1 16:50

import random


class ClassicQuickSort(object):
    """经典快速排序

    只有区分 <= 和 = 两个部分，标志选择为数组固定的一个位置
    """

    def quick_sort(self, arr):
        if arr is None or len(arr) < 2:
            return
        self.quick_sort_main(arr, 0, len(arr) - 1)

    def quick_sort_main(self, arr, l, r):
        """
        快排递归
        :param arr:
        :param l:
        :param r:
        :return:
        """
        if l >= r:
            return
        separate = self.search_swap(arr, l, r)
        self.quick_sort_main(arr, l, separate - 1)
        self.quick_sort_main(arr, separate + 1, r)

    def search_swap(self, arr, l, r):
        """
        根据标志数据的大小，将小的数放置在左边，大的数放置在右边
        :param arr:
        :param l:
        :param r:
        :return:
        """
        flag = r
        while l < r:
            if arr[l] < arr[flag]:
                l = l + 1
            else:
                if arr[r] >= arr[flag]:
                    r = r - 1
                else:
                    arr[l], arr[r] = arr[r], arr[l]
                    l = l + 1
        arr[l], arr[flag] = arr[flag], arr[l]
        return l


class NewQuickSort(object):
    """在经典的快速排序上增加了中间相等的部分，划分为三个区域
    """

    def quick_sort(self, arr):
        if arr is None or len(arr) < 2:
            return
        self.quick_sort_main(arr, 0, len(arr) - 1)

    def quick_sort_main(self, arr, l, r):
        if l >= r:
            return
        left_separates, right_separates = self.search_swap(arr, l, r)
        self.quick_sort_main(arr, l, left_separates - 1)
        self.quick_sort_main(arr, right_separates + 1, r)

    # def search_swap(self, arr, l, r):
    #     num = arr[r]
    #     less = l - 1
    #     more = r + 1
    #     current = l
    #     while current < more:
    #         if arr[current] < num:
    #             less = less + 1
    #             arr[less], arr[current] = arr[current], arr[less]
    #             current = current + 1
    #         elif arr[current] < arr[num]:
    #             more = more -1
    #             arr[more], arr[current] = arr[current], arr[more]
    #         else:
    #             current = current + 1
    #     return less + 1, more - 1

    def search_swap(self, arr, l, r):
        """
        划分为小于等于和大于三个区域
        :param arr:
        :param l:
        :param r:
        :return:
        """
        less = l - 1
        more = r
        while l < more:
            if arr[l] < arr[r]:
                less = less + 1
                arr[less], arr[l] = arr[l], arr[less]
                l = l + 1
            elif arr[l] > arr[r]:
                more = more - 1
                arr[more], arr[l] = arr[l], arr[more]
            else:
                l = l + 1
        arr[more], arr[r] = arr[r], arr[more]
        return less + 1, more


class RandomQuickSort(object):

    def quick_sort(self, arr):
        if arr is None or len(arr) < 2:
            return
        self.quick_sort_main(arr, 0, len(arr) - 1)

    def quick_sort_main(self, arr, l, r):
        """
        增加比较位随机的部分
        :param arr:
        :param l:
        :param r:
        :return:
        """
        if l >= r:
            return
        random_index = random.randint(l, r)
        arr[random_index], arr[r] = arr[r], arr[random_index]
        left_separates, right_separates = self.search_swap(arr, l, r)
        self.quick_sort_main(arr, l, left_separates - 1)
        self.quick_sort_main(arr, right_separates + 1, r)

    def search_swap(self, arr, l, r):
        """划分为小于等于和大于三个区域
        :param arr:
        :param l:
        :param r:
        :return:
        """
        less = l - 1
        more = r
        while l < more:
            if arr[l] < arr[r]:
                less = less + 1
                arr[less], arr[l] = arr[l], arr[less]
                l = l + 1
            elif arr[l] > arr[r]:
                more = more - 1
                arr[more], arr[l] = arr[l], arr[more]
            else:
                l = l + 1
        arr[more], arr[r] = arr[r], arr[more]
        return less + 1, more


def main():
    arr = [2, 3, 1, 5, 4]
    # classic_quick_sort = ClassicQuickSort()
    # classic_quick_sort.quick_sort(arr)
    new_quick_sort = NewQuickSort()
    new_quick_sort.quick_sort(arr)
    for value in arr:
        print(value)


if __name__ == '__main__':
    main()

File: throwcode_java/src/main/java/com/goinhn/algorithm/sort/QuickSort.java
package com.goinhn.algorithm.sort;

/**
 * <p>
 * 快速排序
 * </p>
 * <p>
 * 时间复杂度 O(N*logN)
 * 空间复杂度 O(logN)
 * </P>
 *
 * @author goinhn
 * @date 2020-08-23T23:30:04
 */
public class QuickSort {

    /**
     * 经典快速排序
     * 只有区分 <= 和 = 两个部分，标志选择为数组固定的一个位置
     */
    public static class ClassicQuickSort {

        /**
         * 初始条件判断
         *
         * @param arr
         */
        public static void quickSort(int[] arr) {
            if (arr == null || arr.length < 2) {
                return;
            }

            quickSort(arr, 0, arr.length - 1);
        }

        /**
         * 快速排序
         * 递归进行
         *
         * @param arr
         * @param l
         * @param r
         */
        public static void quickSort(int[] arr, int l, int r) {
            if (l >= r) {
                return;
            }

            int separate = searchSwap(arr, l, r);
            quickSort(arr, l, separate - 1);
            quickSort(arr, separate + 1, r);
        }

        /**
         * 根据标志数据的大小将小的数放置到左边，大的数放置到右边
         *
         * @param arr
         * @param l
         * @param r
         */
        public static int searchSwap(int[] arr, int l, int r) {
            // 设置标志的位置，统一将标志位设置为数组的最后一个数
            int flag = r;
            while (l < r) {
                if (arr[l] < arr[flag]) {
                    l++;
                } else {
                    if (arr[r] >= arr[flag]) {
                        r--;
                    } else {
                        swap(arr, l++, r);
                    }
                }
            }

            swap(arr, l, flag);
            return l;
        }

        /**
         * 交换方法
         *
         * @param arr
         * @param x
         * @param y
         */
        public static void swap(int[] arr, int x, int y) {
            int temp = arr[x];
            arr[x] = arr[y];
            arr[y] = temp;
        }

        public static void main(String[] args) {
            int[] arr = new int[]{1, 1, 1, 1, 2, 1, 2};
            quickSort(arr);
            for (int i : arr) {
                System.out.println(i);
            }
        }

    }


    /**
     * 在经典的快速排序上增加了中间相等的部分，划分为三个区域
     */
    public static class NewQuickSort {

        /**
         * 初始条件判断
         *
         * @param arr
         */
        public static void quickSort(int[] arr) {
            if (arr == null || arr.length < 2) {
                return;
            }

            quickSort(arr, 0, arr.length - 1);
        }

        /**
         * 快速排序
         * 递归进行
         *
         * @param arr
         * @param l
         * @param r
         */
        public static void quickSort(int[] arr, int l, int r) {
            if (l >= r) {
                return;
            }

            // 中间相等的部分的分隔，第一个数为相等的左侧边界，第二个数为右侧相等的部分
            int[] equalSeparates = searchSwap(arr, l, r);
            quickSort(arr, l, equalSeparates[0] - 1);
            quickSort(arr, equalSeparates[1] + 1, r);
        }

        /**
         * 根据标志数据的大小将小的数放置到左边，大的数放置到右边
         *
         * @param arr
         * @param l
         * @param r
         */
//        public static int[] searchSwap(int[] arr, int l, int r) {
//            // 设置标志的位置，统一将标志位设置为数组的最后一个数
//            int num = arr[r];
//            int less = l - 1;
//            int more = r + 1;
//            int current = l;
//
//            while (current < more) {
//                if (arr[current] < num) {
//                    swap(arr, ++less, current++);
//                } else if (arr[current] > num) {
//                    swap(arr, --more, current);
//                } else {
//                    current++;
//                }
//            }
//
//            return new int[]{less + 1, more - 1};
//        }
        public static int[] searchSwap(int[] arr, int l, int r) {
            int less = l - 1;
            int more = r;
            while (l < more) {
                if (arr[l] < arr[r]) {
                    swap(arr, ++less, l++);
                } else if (arr[l] > arr[r]) {
                    swap(arr, --more, l);
                } else {
                    l++;
                }
            }
            swap(arr, more, r);
            return new int[]{less + 1, more};
        }

        /**
         * 交换方法
         *
         * @param arr
         * @param x
         * @param y
         */
        public static void swap(int[] arr, int x, int y) {
            int temp = arr[x];
            arr[x] = arr[y];
            arr[y] = temp;
        }

        public static void main(String[] args) {
            int[] arr = new int[]{1, 2, 3, 1, 5, 6, 3, 2, 1, 0};
            quickSort(arr);
            for (int i : arr) {
                System.out.println(i);
            }
        }

    }

    /**
     * 随机快速排序，标志位的选择为随机，符合概率分布
     */
    public static class RandomQuickSort {

        /**
         * 初始条件判断
         *
         * @param arr
         */
        public static void quickSort(int[] arr) {
            if (arr == null || arr.length < 2) {
                return;
            }

            quickSort(arr, 0, arr.length - 1);
        }

        /**
         * 快速排序
         * 递归进行
         *
         * @param arr
         * @param l
         * @param r
         */
        public static void quickSort(int[] arr, int l, int r) {
            if (l >= r) {
                return;
            }

            // 随机选择标志位，和最后一个位置的数进行对换
            swap(arr, r, (int) (Math.random() * (r - l + 1)) + l);
            int[] equalSeparates = searchSwap(arr, l, r);
            quickSort(arr, l, equalSeparates[0] - 1);
            quickSort(arr, equalSeparates[1] + 1, r);
        }

        /**
         * 根据标志数据的大小将小的数放置到左边，大的数放置到右边
         *
         * @param arr
         * @param l
         * @param r
         */
//        public static int[] searchSwap(int[] arr, int l, int r) {
//            // 设置标志的位置，统一将标志位设置为数组的最后一个数
//            int num = arr[r];
//            int less = l - 1;
//            int more = r;
//            int current = l;
//
//            while (current < more) {
//                if (arr[current] < num) {
//                    swap(arr, ++less, current++);
//                } else if (arr[current] > num) {
//                    swap(arr, --more, current);
//                } else {
//                    current++;
//                }
//            }
//
//            return new int[]{less + 1, more - 1};
//        }
        public static int[] searchSwap(int[] arr, int l, int r) {
            int less = l - 1;
            int more = r;
            while (l < more) {
                if (arr[l] < arr[r]) {
                    swap(arr, ++less, l++);
                } else if (arr[l] > arr[r]) {
                    swap(arr, --more, l);
                } else {
                    l++;
                }
            }
            swap(arr, more, r);
            return new int[]{less + 1, more};
        }

        /**
         * 交换方法
         *
         * @param arr
         * @param x
         * @param y
         */
        public static void swap(int[] arr, int x, int y) {
            int temp = arr[x];
            arr[x] = arr[y];
            arr[y] = temp;
        }

        public static void main(String[] args) {
            int[] arr = new int[]{1, 2, 3, 1, 5, 6, 3, 2, 1, 0};
            quickSort(arr);
            for (int i : arr) {
                System.out.println(i);
            }
        }
    }
}
